### PR TITLE
Bug 1875187 - Avoid race condition setting mode before opening a tab

### PR DESCRIPTION
**Pull request by MatthewTighe**
### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1875187

### CODE REVIEW

**Comment by sv-ohorvath** [comment]
@MatthewTighe I saw today this test which seems to be affected by the same issue. Can you take a look, please? https://console.firebase.google.com/project/moz-fenix/testlab/histories/bh.66b7091e15d53d45/matrices/6854215613285633784/executions/bs.9ca07fa2ef429fc4

**Comment by MatthewTighe** [comment]
The `RecentlyClosed` test definitely seems related, but I don't think the `Bookmarks` test follows a codepath I'd expect to be problematic. I'll have to look into it more.

**Comment by MatthewTighe** [comment]
@sv-ohorvath Still not sure that the bookmarks test is related. It looks like that test is failing to find the 3 dot menu in the toolbar. Is that right?

In the meantime, I've updated the patch to fix the bugs for top sites and recently closed. My hope is to land it before the cut next week so we can avoid these bugs in release.

File: fenix/app/src/main/java/org/mozilla/fenix/home/intent/ReEngagementIntentProcessor.kt [hunk 1]
```diff
@@ -49,12 +46,12 @@ class ReEngagementIntentProcessor(
                         navController.nav(null, directions, options)
                     }
                     else -> {
-                        appStore.dispatch(AppAction.OpenTabInBrowser(BrowsingMode.Private))
```
Comment (MatthewTighe) [inline fenix/app/src/main/java/org/mozilla/fenix/home/intent/ReEngagementIntentProcessor.kt:52]: This was the only other place using this, so I think it makes sense to just remove it now that we have the ability to

Comment (gabrielluong) [reply fenix/app/src/main/java/org/mozilla/fenix/home/intent/ReEngagementIntentProcessor.kt:52]: I think we need to also remove this in https://searchfox.org/mozilla-mobile/source/firefox-android/fenix/app/src/test/java/org/mozilla/fenix/home/intent/ReEngagementIntentProcessorTest.kt#89

**Comment by gabrielluong** [comment]
@MatthewTighe I was ready to approve this PR, but seeing that we are still uncertain about the bookmark tests. I am thinking we should instead back out the regressing changes and reland after the cut happens. There is really no time to verify that your changes here has solve all the lingering issues if we landed this today since it is soft freeze. I think it would be wise to avoid any uncertainty and risk and just buy ourselves a bit more time.

**Comment by MatthewTighe** [comment]
Original patch reverted, but I am still curious for more details about the bookmarks test case failure if you are able to provide them @sv-ohorvath. We will plan to address this patch and the bookmarks bug after the start of the new nightly cycle next week

**Comment by gabrielluong** [comment]
@MatthewTighe Do you think we should close this PR?

**Comment by MatthewTighe** [comment]
Yeah, I'll close this for now and include it in the re-ordered patches for https://bugzilla.mozilla.org/show_bug.cgi?id=1861459

**Pull request closed without merging**